hathor/transaction/storage/transaction_storage.py

```diff
@@ -988,6 +998,58 @@ def _topological_fast(self) -> Iterator[BaseTransaction]:
         yield from cur_blocks
         yield from cur_txs
 
+    def _topological_fast2(self) -> Iterator[BaseTransaction]:
```

Comment:
I feel the following implementation should work:

```python
to_visit: List[Item] = list(map(Item, self.get_all_genesis()))
seen: Set[bytes] = set()
heapq.heapify(to_visit)
while to_visit:
    item = heapq.heappop(to_visit)
    yield item.tx
    # We can safely discard because no other tx will try to visit this one.
    seen.discard(item.tx.hash)
    for child_tx_hash in item.tx.get_metadata().children:
        if child_tx_hash in seen:
            continue
        child_tx = self.get_transaction(child_tx_hash)
        heapq.heappush(to_visit, Item(child_tx))
        seen.add(child_tx_hash)
```

Note that both my implementation and yours assume that for all tx there exists a path between genesis and that tx. In other words, if a transaction is disconnected, it will not be yielded.

Reply:
Your assumption is part of the protocol, so I'd say it's an implied assumption as much as the structure of the transaction (parents, inputs, metadata, etc).

I think your implementation is correct, I had a somewhat similar previous implementation (but using `sortedcontainers.SortedKeyList` instead of `list+heapq`) and I had some issue with yielding transactions not in timestamp order, which I didn't investigate too well and went with an implementation based off the indexes one (there it makes more sense to have a cur_timestamp). I think your proposal is much better than what's currently implemented, I'll use that and test if everything is OK.

Reply:
Done.

It seems to be working great and it's much simpler! I'll test if there's any performance gained (which probably there won't be because complexity is the same and rocksdb reads are the bottleneck).

Reply:
Great! Thanks for testing it.